source/includes/write/bulk-write.cpp

```diff
@@ -0,0 +1,141 @@
+                                        kvp("restaurant_id", "1234"));
+        
+        mongocxx::model::insert_one insert_op{insert_doc.view()};
+        bulk.append(insert_op);
+        // end-bulk-insert-one
+    }
+
+    {
+        // Creates a write model to specify an update one operation and adds it to the bulk operation
+        // start-bulk-update-one
+        auto filter_doc = make_document(kvp("name", "Mongo's Deli"));
+        auto update_doc = make_document(kvp("$set", make_document(kvp("cuisine", "Sandwiches and Salads"))));
+        
+        mongocxx::model::update_one update_op{filter_doc.view(), update_doc.view()};
+        bulk.append(update_op);
+        // end-bulk-update-one
+    }
+
+    {
+        // Creates a write model to specify an update many operation and adds it to the bulk operation
+        // start-bulk-update-many
+        auto filter_doc = make_document(kvp("name", "Mongo's Deli"));
+        auto update_doc = make_document(kvp("$set", make_document(kvp("cuisine", "Sandwiches and Salads"))));
+        
+        mongocxx::model::update_many update_op{filter_doc.view(), update_doc.view()};
+        bulk.append(update_op);
+        // end-bulk-update-many
+    }
+
+    {
+        // Creates a write model to specify a replace operation and adds it to the bulk operation
+        // start-bulk-replace-one
+        auto filter_doc = make_document(kvp("restaurant_id", "1234"));
+        auto replace_doc = make_document(kvp("name", "Mongo's Pizza"),
+                                        kvp("cuisine", "Pizza"),
+                                        kvp("borough", "Brooklyn"),
+                                        kvp("restaurant_id", "5678"));
+        
+        mongocxx::model::replace_one replace_op{filter_doc.view(), replace_doc.view()};
+        bulk.append(replace_op);
+        // end-bulk-replace-one
+    }
+
+    {
+        // Creates a write model to specify a delete one operation and adds it to the bulk operation
+        // start-bulk-delete-one
+        auto filter_doc = make_document(kvp("restaurant_id", "5678"));
+        
+        mongocxx::model::delete_one delete_op{filter_doc.view()};
+        bulk.append(delete_op);
+        // end-bulk-delete-one
+    }
+
+    {
+        // Creates a write model to specify a delete many operation and adds it to the bulk operation
+        // start-bulk-delete-many
+        auto filter_doc = make_document(kvp("name", "Mongo's Deli"));
+        
+        mongocxx::model::delete_many delete_op{filter_doc.view()};
+        bulk.append(delete_op);
+        // end-bulk-delete-many
+    }
+
+    {
+        // Runs the bulk operation based on the instructions in each write model
+        // start-bulk-run
+        auto bulk = collection.create_bulk_write();
+
+        // Specifies documents to insert, update, replace, or delete
+        auto insert_doc = make_document(kvp("name", "Mongo's Deli"),
```

Comment:
C: I'm a little confused on why in this bulk operation, we're creating a new "Mongo's Deli" document and updating it with a new cuisine value, just to then replace it with a Mongo's Pizza" document. I'm also confused on why we are deleting all Mongo's Deli documents when we just replaced a Mongo's Deli document. 

Maybe instead of replacing Mongo's Deli with Mongo's Pizza, we can replace Mongo's Deli in Manhattan with Mongo's Deli in Brooklyn, and then for the delete filter we could delete all restaurants in Manhattan?

Reply:
Sure, that works! updated